File: app/src/main/java/com/abelcht/n01f2smwc/smartwatch/communication/packages/ChangeUVTemperatureAltitudeBarometerPackage.kt
package com.abelcht.n01f2smwc.smartwatch.communication.packages;

class ChangeUVTemperatureAltitudeBarometerPackage(
    uv: Int, temperature: Int, altitude: Int, barometer: Int
) : SmartWatchPackage(
    0x1b.toByte(),
    arrayOf(
        (barometer % 256).toByte(), ((barometer / 256) % 256).toByte(),
        ((barometer / (256 * 256)) % 256).toByte(), 0x00.toByte(), // Barometer
        (altitude % 256).toByte(), ((altitude / 256) % 256).toByte(), // Altitude
        ((temperature * 10) % 256).toByte(),
        (((temperature * 10) / 256) % 256).toByte(), // Temperature
        (if (uv in 1..5) uv else 6).toByte() // UV
    )
)

File: python_framework/packages/ChangeUvTemperatureAltitudeBarometerPackage.py
from datetime import datetime

from packages.SmartWatchPackage import SmartWatchPackage


class ChangeUvTemperatureAltitudeBarometerPackage(SmartWatchPackage):
    def __init__(self, uv: int, temperature: int, altitude: int, barometer: int):
        # Set temperature
        if 0 <= temperature <= 99:
            temperature_fields = [(temperature * 10) % 256, int((temperature * 10) / 256) % 256]
        else:
            temperature_fields = [0x00, 0x00]

        # Set UV
        if 1 <= uv <= 5:
            uv_fields = [uv]
        elif uv == 0:
            # Looks like a workaround
            uv_fields = [0x06]
        else:
            uv_fields = [0x01]

        # Set altitude
        if 0 <= altitude <= 9999:
            altitude_fields = [altitude % 256, int(altitude / 256) % 256]
        else:
            altitude_fields = [0x00, 0x00]

        # Set barometer
        if 0 <= barometer <= 299999:
            barometer_fields = [barometer % 256, int(barometer / 256) % 256, int(barometer / (256 ** 2)) % 256, 0x00]
        else:
            barometer_fields = [0x00, 0x00, 0x00, 0x00]

        message_type = 0x1b
        content = barometer_fields + altitude_fields + temperature_fields + uv_fields
        super().__init__(message_type, content)
